src/sentry/static/sentry/app/components/hovercard.jsx

```diff
@@ -70,6 +79,71 @@ class Hovercard extends React.Component {
     return currentClasses.length > 0 ? currentClasses : classes;
   }
 
+  // figure out if we're dealing with hovercard-left or hovercard-right (in which case we don't
+  // want to center the hovercard) or hovercard or hovercard-bottom (in which case we do)
+  getMargin(classes) {
+    if (!this.cardElement || !this.state.visible) return {};
+
+    const rect = this.cardElement.getBoundingClientRect();
+
+    const rectWidth = this.isMultipleLines() ? rect.width : this.getSingleLineWidth();
+
+    if (classes.indexOf('left') !== -1 || classes.indexOf('right')) {
+      return -rectWidth / 2;
+    } else {
+      return 0;
+    }
+  }
+
+  isMultipleLines() {
+    if (!this.cardElement || !this.state.visible) return {};
+
+    // spans will create a new bounding box for each line of text
+    // if the text wraps, so this is a way to test for text wrapping
+    return this.cardBodySpan.getClientRects().length > 1;
```

Comment:
this is really cool, and should be all the code you need. just apply a prop, something like

`<CardContainer multiLine={this.isMultipleLines}>`

and then in the css:

`
width: ${p => p.multiLine ? whateverWidthIsBest : "auto"};
`

Reply:
Hmmm - not sure if I understand totally. Should I create a new component (like CardContainer) to live inside the Hovercard? Should it be a styled component or a real one?

Also, the trick is, in order for the `this.cardBodySpan.getClientRects().length > 1` bit to work, the text has to first render (invisibly) in the fixed-width container. I _think_ that works given what's currently there, but relies on the default min-width being there.

Finally, the `whateverWidthIsBest` bit is the hard part - that's what I was getting using all of that (admittedly very brittle) style-checking. The only thing which shrinks to fit the text on the first pass is the inner span, but we're then setting the width on the outer container/div... which is why one then has to take into account all the included padding, margins, etc.

(All of the above said as well, it's occurring to me that all of this is predicated on the idea of it specifically being _text_ and specifically being _text in the body_, which isn't guaranteed... :-/ )